[install README CentOS and Ubuntu] 21 install updates, including make - bugfix (#5203)
aces/Loris: README.CentOS7.md

# LORIS CentOS 7.x Notes

This document contains supplemental details on how to perform a basic CentOS 7.x install of LORIS.
Note that the main README in LORIS assumes that LORIS is being run on Ubuntu.

This Readme assumes you already understand basic UNIX, MySQL and Apache setup and
settings. If you're not already comfortable troubleshooting sysadmin issues,
you should not follow this guide.

For further details on the install process, please see the LORIS GitHub Wiki CentOS Install page.  

# System Requirements - Install dependencies

Default dependencies installed by CentOS 7.x may not meet the version requirements for LORIS deployment or development:
* MariaDB 10.3 is supported for LORIS 21.* 
* PHP 7.2 is supported for LORIS 21.*

In addition to the above, the following packages should be installed with `yum` and may also differ from the packages referenced in the main (Ubuntu) [LORIS Readme](./README.md). Detailed command examples are provided below.

>  * Apache2

Reviewer: ```suggestion
 * Apache2.4
```

 * Node

>  * PHP Composer

Reviewer 2: Should this composer be removed here, @kongtiaowang ?

Author: no, we need composer here.

>  * PHP Composer

Reviewer 3: What step in the install requires sudo?

>  * PHP Composer

Reviewer: chg, chmod (sometime), yum install, systemctl

>  * PHP Composer

Reviewer: Can this be added? It correspond to Step 1 of README.md file adapted for CentOS7

>  * PHP Composer

Reviewer: @kongtiaowang This should be added at the beginning (before any install step).
It is the equivalent of Step 1 of main README.md file for CentOS7.  Step 1 as done in README.md file will not work on CentOS

>  * PHP Composer

Reviewer 3: Yeah I'd prefer it just go in at this point since a bugfix release is needed.

sudo yum update
sudo yum install php72
sudo yum install php72-php-fpm php72-php-gd php72-php-json php72-php-mbstring php72-php-mysqlnd php72-php-xml php72-php-xmlrpc php72-php-opcache php72-php-pdo php72-php-mysql
```
## MariaDB

MySQL is not recommended/supported on CentOS, see [paragraph 2 here](https://www.digitalocean.com/community/tutorials/how-to-install-mysql-on-centos-7) for background). MySQL and MariaDB are nearly identical in operation.

*Note:* LORIS developers (those NOT working with a .zip release codebase) should skip steps relating to hosting their database locally. Contact your sysadmin for database credentials.

```
sudo yum install MariaDB-client MariaDB-server
```

By default in CentOS 7, the MariaDB version is (currently) 10.2.27
Check what version you have installed by running:
```
mysql -v
```

Upgrade your MariaDB to version 10.3 (LORIS 21 does not support MariaDB 10.2.27)
Then, 
```
sudo service mariadb start
sudo service mariadb status
```

To finalise the MySQL/MariaDB installation: 
```
mysql_secure_installation
```
Then follow instructions to create a password for the root user.

## NodeJS
```
sudo yum install nodejs
```

> ## PHP Composer

Reviewer 2: remove this section? @kongtiaowang

Reviewer 3: Composer is still needed, it's just run by a script within `make`. So these sections are needed.

```
sudo curl -sS https://getcomposer.org/installer | php
sudo mv composer.phar /usr/local/bin/composer
```
# Get the Code

Download the latest release from the [releases page](https://github.com/aces/Loris/releases) to the home directory (~/), unzip it, and copy the contents to your project directory, `/var/www/loris` (we recommend naming your project directory `loris`, although you can use a different naming convention if you prefer). 
```

> wget https://github.com/aces/Loris/archive/v$VERSION.zip

Reviewer 3: ```suggestion
curl -s https://api.github.com/repos/aces/loris/releases/latest \
| grep 'tarball' \
| cut -d : -f2,3 \
| tr -d \" \
| tr -d , \
| wget -i -
```

This command will use the GitHub API to always get the latest release. It might be useful but you don't have to include it.

unzip Loris-%VERSION%.zip

> cp -r Loris-%VERSION%/* /var/www/loris

Reviewer 3: ```suggestion
cp -r Loris-%VERSION%/ /var/www/loris
```

`-r` will take care of copying the files within the directory.

 * DocumentRoot should point to `/var/www/loris/htdocs`
 * The `smarty/templates_c/` directory must be writable by Apache (e.g. by running: `chmod 775 /var/www/loris/smarty/templates_c)

Finally, restart apache:
```
sudo service httpd restart
sudo service httpd status
```
# Install LORIS

For the purpose of following LORIS conventions and easy understanding of all LORIS documentation, we recommend the following account names: 

* `lorisadmin` unix user with sudo permission who will setup and manage the LORIS back-end
* `lorisuser` MySQL database user with limited (insert, delete, update, select...) permissions, for database transactions requested by the Loris front-end
* `admin` default front-end username for LORIS administrator account (login via browser)

## Run the LORIS install script to set up directories

```
cd /var/www/loris/tools
./install.sh
```
Run the makefile (use `make dev` if you are setting up a development sandbox)
```

> cd /var/www/$projectname

Reviewer 3: These instructions should result in it being `/var/www/loris` via line 83. Also the instructions just above specify `loris` so this should stay consistent.

make
```
## Install your database
Open your browser and go to: `http://%IPADDRESS%/installdb.php`

(%IPADDRESS% will likely be the IP address of the VM you are ssh'ed into)

The web page will prompt you for the following information: 

 * `Server Hostname` localhost if your database if hosted on your VM or the IP address of your database server

 * `Admin Username` should be `root` or a unix user with a high level of privilege on your VM

>  * `Admin Password` whatever password you set when running the `mysql_secure_installation` step

Reviewer: ```suggestion
 * `Admin Password` The password you set for `root` user when running the `mysql_secure_installation` step
```

 * `Database Name` set to "LORIS" by default, can be customised

> Click submit, and on the next screen that is presented, follow instructions to enter the username and password of your backend user (`lorisuser`), and front-end `admin` user. 

Reviewer: ```suggestion
Click submit, and on the next screen that is presented, follow instructions to enter the username and password of your LORIS database user (`lorisuser`), and front-end `admin` user. 
```

If you encounter issues creating/generating your config file, you may have to manually paste the xml output that appears on the screen into the file `/var/www/loris/project/config.xml`

Your LORIS instance should now be accessible by pointing your browser URL to `http://%IPADDRESS%`

If there are any errors or you get a blank page, troubleshoot the errors in your apache error log - by default 
 `/var/log/httpd/loris-error.log`

**Next:** Follow the [**Setup Guide** in the LORIS wiki](https://github.com/aces/Loris/wiki/Setup) for all post-install steps and troubleshooting.  

> Config settings can be accessed via the front-end Config module under the Admin drop-down menu.

Reviewer: SHould not it have the same `Community`, `Contributing` and `Powered by MCIN` section as main README.md?

Reviewer 3: I don't think it needs these fields. The only reason `README.md` has this is because it's doing double duty as a landing page for the repo.

# LORIS CentOS 7.x Notes

This document contains supplemental details on how to perform a basic CentOS 7.x install of LORIS.
Note that the main README in LORIS assumes that LORIS is being run on Ubuntu.

This Readme assumes you already understand basic UNIX, MariaDB (MySQL) and Apache setup and
settings. If you're not already comfortable troubleshooting system administration issues,

> you should not follow this guide.

Reviewer: I would replace the last phrase by something more like
These step should be done by someone with Linux command line knowledge.
Currently it is written to scare people away.

Reviewer 2: Great. just made it slightly less scary in [commit 49320e](https://github.com/aces/Loris/pull/5203/commits/49320e31d1884ae975eebed30f66a7e2399622f1) - could you verify/re-review @PapillonMcGill ? - thx

Reviewer: @christinerogers Way better, I just requested a little change to current text. (see below).

For further details on the install process, please see the LORIS GitHub Wiki CentOS Install page.  

# System Requirements - Install dependencies

Default dependencies installed by CentOS 7.x may not meet the version requirements for LORIS deployment or development:

> * MariaDB 10.2 is supported for LORIS 21.* 

Reviewer: ```suggestion
* MariaDB 10.3 is supported for LORIS 21.* 
```
as indicated by @driusan

Reviewer 2: thanks @PapillonMcGill we'll fix this typo

Reviewer: @christinerogers Ok!

sudo yum install http://rpms.remirepo.net/enterprise/remi-release-7.rpm
sudo yum install yum-utils
sudo yum-config-manager --enable remi-php72
sudo yum update
sudo yum install php72
sudo yum install php72-php-fpm php72-php-gd php72-php-json php72-php-mbstring php72-php-mysqlnd php72-php-xml php72-php-xmlrpc php72-php-opcache php72-php-pdo php72-php-mysql
```
## MariaDB

MySQL is not recommended/supported on CentOS, see [paragraph 2 here](https://www.digitalocean.com/community/tutorials/how-to-install-mysql-on-centos-7) for background). MySQL and MariaDB are nearly identical in operation.

*Note:* LORIS developers (those NOT working with a .zip release codebase) should skip steps relating to hosting their database locally. Contact your sysadmin for database credentials.

```
sudo yum install MariaDB-client MariaDB-server
```

By default in CentOS 7, the MariaDB version is 10.2
Check what version you have installed by running:
```
mysql -v
```

Upgrade your MariaDB to version 10.3 (LORIS 21 does not support MariaDB 10.2.27)
Then, 
```
sudo systemctl start mariadb 
sudo systemctl enable mariadb
sudo service mariadb status
```

To finalise the MySQL/MariaDB installation: 
```

> mysql_secure_installation

Reviewer: ```suggestion
sudo mysql_secure_installation
```

# LORIS CentOS 7.x Notes

This document contains supplemental details on how to perform a basic CentOS 7.x install of LORIS.
Note that the main README in LORIS assumes that LORIS is being run on Ubuntu.

> This Readme assumes some familiarity with UNIX, MariaDB (MySQL) and Apache.

Reviewer: ```suggestion
This Readme assumes some familiarity with Linux, MariaDB (MySQL) and Apache.
```

For further details on the install process, please see the LORIS GitHub Wiki CentOS Install page.  

# System Requirements - Install dependencies

Default dependencies installed by CentOS 7.x may not meet the version requirements for LORIS deployment or development:
* MariaDB 10.3 is supported for LORIS 21.* 
* PHP 7.2 is supported for LORIS 21.*

In addition to the above, the following packages should be installed with `yum` and may also differ from the packages referenced in the main (Ubuntu) [LORIS Readme](./README.md). Detailed command examples are provided below (`sudo` privilege may be required depending on your system).
 * Apache2.4
 * [Composer](https://getcomposer.org)
 * Node

>  * NPM

Reviewer 3: Could you specify that the version must be >= 8.0?

 * make

## Apache2

> ```

Reviewer 3: ```suggestion
```bash
```

sudo yum install httpd 
sudo systemctl enable httpd
sudo systemctl start httpd
```
## PHP 7.2

> ```

Reviewer 3: ```suggestion
```bash
```

sudo yum install epel-release
sudo yum install http://rpms.remirepo.net/enterprise/remi-release-7.rpm
sudo yum install yum-utils
sudo yum-config-manager --enable remi-php72
sudo yum update
sudo yum install php72
sudo yum install php72-php-fpm php72-php-gd php72-php-json php72-php-mbstring php72-php-mysqlnd php72-php-xml php72-php-xmlrpc php72-php-opcache php72-php-pdo php72-php-mysql
```
## MariaDB

MySQL is not recommended/supported on CentOS, see [paragraph 2 here](https://www.digitalocean.com/community/tutorials/how-to-install-mysql-on-centos-7) for background). MySQL and MariaDB are nearly identical in operation.

*Note:* LORIS developers (those NOT working with a .zip release codebase) should skip steps relating to hosting their database locally. Contact your system administrator for database credentials.

> ```

Reviewer 3: ```suggestion
```bash
```

sudo yum install MariaDB-client MariaDB-server
```

By default in CentOS 7, the MariaDB version is 10.2
Check what version you have installed by running:
```
mysql -V
```

Upgrade your MariaDB to version 10.3 (LORIS 21 does not support MariaDB 10.2.27)
Then, 

> ```

Reviewer 3: ```suggestion
```bash
```

sudo systemctl start mariadb 
sudo systemctl enable mariadb
sudo systemctl status mariadb
```

To finalise the MariaDB installation: 

> ```

Reviewer 3: ```suggestion
```bash
```

sudo mysql_secure_installation
```
Then follow instructions to create a password for the root user.

## NodeJS

> ```

Reviewer 3: ```suggestion
```bash
```

sudo yum install nodejs
```

## PHP Composer

> ```

Reviewer 3: ```suggestion
```bash
```

sudo curl -sS https://getcomposer.org/installer | php
sudo mv composer.phar /usr/local/bin/composer
```
# Get the Code

Download the latest release from the [releases page](https://github.com/aces/Loris/releases) to the home directory (~/), unzip it, and copy the contents to your project directory, `/var/www/loris` (we recommend naming your project directory `loris`, although you can use a different naming convention if you prefer). 

> ```

Reviewer 3: ```suggestion
```bash
```

# Configure Apache

A sample apache configuration file is in `docs/config/apache2-site`. 
The install script will ask if you want to automatically create/install apache config files.
It is recommended to perform this step **manually** --: copy our sample file to the apache configuration directory (`/etc/httpd/conf.d/`) and add the `.conf` file extension:

```bash
cd /var/www/loris
cp docs/config/apache-site /etc/httpd/conf.d/apache-site.conf
```

Customize and Verify your settings: 
* Paths and settings in `/etc/httpd/conf.d/apache-site.conf` should be populated appropriately for your server. Replace placeholders such as `%LORISROOT%` with `/var/www/loris`, `%PROJECTNAME%` with `loris`, `%LOGDIRECTORY%` with `/var/log/httpd/loris-error.log` 
 * DocumentRoot should point to `/var/www/loris/htdocs`
 * The `smarty/templates_c/` directory must be writable by Apache (e.g. by running: `sudo chgrp -R httpd /var/www/loris/smarty/templates_c` and `sudo chmod 775 /var/www/loris/smarty/templates_c`).

Finally, restart apache:

> ```

Reviewer 3: ```suggestion
```bash
```

sudo systemctl restart httpd
sudo systemctl status httpd
```
# Install LORIS

For the purpose of following LORIS conventions and easy understanding of all LORIS documentation, we recommend the following account names: 

> * `lorisadmin` unix user with sudo permission who will setup and manage the LORIS back-end

Reviewer: If we recommend `lorisadmin` not to have sudo permission, this line should be modified

Reviewer 2: Hi @PapillonMcGill  I think this inconsistency/irregularity has to be smoothed out via discussion per John's new issue #5297 on this topic - so we have clarity and consensus across the team. 
Definitely noted; at the same time, getting this PR merged quickly on `bugfix` for 21.0.3, to correct for the grosser inaccuracies in the major release documentation, would be ideal (whereas sudo is a clarification of our design/practices going forward).

* `lorisuser` MySQL database user with limited (insert, delete, update, select...) permissions, for database transactions requested by the Loris front-end
* `admin` default front-end username for LORIS administrator account (login via browser)

## Run the LORIS install script to set up directories

> ```

Reviewer 3: ```suggestion
```bash
```

cd /var/www/loris/tools
./install.sh
```
Run the makefile (use `make dev` if you are setting up a development sandbox)

> ```

Reviewer 3: ```suggestion
```bash
```

cd /var/www/loris
make
```
## Install your database
Open your browser and go to: `http://%IPADDRESS%/installdb.php`

(%IPADDRESS% will likely be the IP address of the VM you are ssh'ed into)

The web page will prompt you for the following information: 

>  * `Server Hostname` localhost if your database if hosted on your VM or the IP address of your database server

Reviewer 3: ```suggestion
 * `Server Hostname` localhost if your database is hosted on your VM or the IP address of your database server
```

Reviewer 2: Thanks @johnsaigle -  I just fixed this typo in [commit b2080db](https://github.com/aces/Loris/pull/5203/commits/b2080dbb39694f974074dd834b271df6b1be1c72) - could you verify/resolve?

# System Requirements

Default dependencies installed by CentOS 7.x may not meet the version requirements LORIS deployment or development.

> * MySQL 5.7 is supported for LORIS 21.*

Reviewer: should not that MariaDB (10.?) is also supported as it is the default for CentOS 7
@ridz1208 which version of MariaDB is the minimum?

Reviewer 2: Dave says on slack - 10.3

```
**PHP:**
```
sudo yum install epel-release
sudo yum install http://rpms.remirepo.net/enterprise/remi-release-7.rpm
sudo yum install yum-utils
sudo yum-config-manager --enable remi-php72
sudo yum update
sudo yum install php72

> sudo yum install php72-php-fpm php72-php-gd php72-php-json php72-php-mbstring php72-php-mysqlnd php72-php-xml php72-php-xmlrpc php72-php-opcache php72-php-pdo php72-php-mysql

Reviewer: @christinerogers Nope, all needed for production. Except maybe php72-php-opcache, needed to accelerate apache response by caching code.

**MySQL:**

*Note:* Loris developers (those NOT working with a .zip release codebase) should skip steps relating to hosting mysql locally. Contact your sysadmins for database credentials directly.

Running the following command will donnload and install MariaDB by default in CentOS 7 - for details: [(e.g. Tutorial, see paragraph2)](https://www.digitalocean.com/community/tutorials/how-to-install-mysql-on-centos-7).

``` 

> sudo yum install mysql mysql-server

Reviewer 2: Thanks @PapillonMcGill  I'll make the commit (cc @kongtiaowang

> sudo yum install mysql mysql-server

Reviewer 2: ... i was wondering about that because of the default mariadb installation it already mentioned.  so, extra good catch @PapillonMcGill  👍

> sudo yum install mysql mysql-server

Reviewer: CentOS 8 release 2 days ago come with MariaDB 10.3.
Should we instead modify this readme for CentOS8?

> sudo yum install mysql mysql-server

Reviewer 2: I don't think we can test on CentOS 8 yet, and can't assume others will be using it yet. 
Should we just say `CentOS 7 or higher`?

> sudo yum install mysql mysql-server

Reviewer: I can build a VM with CentOS8 and try it. Or at least make it accessible so you can try it.
Can't promise a time frame but hopefully by the end of the week.

> sudo yum install mysql mysql-server

Reviewer: @kongtiaowang I would just name this file READMECentOS.md  no version number in the file and put a not at the top that it apply to CentOS version 7.

cp -r Loris-%VERSION%/* /var/www/loris
```

Alternatively the latest development branch can be obtained by forking the [LORIS repository](http://github.com/aces/Loris) for development purposes. We do not support unstable dev branches. 

# Setup

## Composer

Composer will download all of LORIS's library requirements, assuming an active internet connection.
This must be done from the LORIS project directory `/var/www/loris`. There may be additional packages
to install at this setp for composer to exist successfully, be sure to install the php 7 compatible 
versions of these packages, if necessary. 

**Note** It may be necessary to create the `project/libraries/` directory before running _composer install_ 

```bash

> composer install --no-dev

Author: Not, make includes the "composer install " command.
